### ISSUE TITLE
Format of the second argument of AddFinalDerivation

### ISSUE THREAD

**Issue by zickgraf** [homalg-project/CAP_project#590]
I could not find any documentation on the format of the second argument of `AddFinalDerivation`.

From looking at existing final derivations, it should be a list of pairs `[ operation, number_of_occurrences ]`. If this is correct, the second argument of the final derivation of `BasisOfExternalHom` is missing the entries `number_of_occurrences`.

Note: this might get relevant when fixing #588.

**Comment by sebastianpos** [comment]
`BasisOfExternalHom` uses the experimental feature of a "bundle of final derivations", which still is not implemented in a clean way, see #382.

The idea is to install not only a single but a bundle of functions provided that a set of conditions is satisfied after the finalize step. These conditions consist of a list of basic operations that have to exist and a list of basic operations that are not allowed to exist.

In the case of `BasisOfExternalHom`, the second function that comes along in this bundle is `CoefficientsOfMorphismWithGivenBasisOfExternalHom`.

I think `BasisOfExternalHom` having weight 1 now comes from the unfortunate issue that the current "interface" for bundle final derivations does not comply with the `ConditionsListComplete` option, in the sense that `ConditionsListComplete` relies on the 2nd argument in `AddFinalDerivation` being a list of functions and multiplicities that occur in the function to be installed, whereas `AddFinalDerivation` in the bundle case treats the second argument merely as part of the condition when to install the whole bundle.

**Comment by zickgraf** [comment]
Thanks for the detailed explanation!

> These conditions consist of a list of basic operations that have to exist [...]

Okay, currently it is implemented as a list of lists which have the basic operations as their first entries.

While looking at existing final derivations in Freyd, I have noticed another possible problem: `WeakKernelObject` is handled as a bundle but `WeakCokernelObject`, `WeakBiFiberProductMorphismToDirectSum`, etc. are not.

I was not aware of #382. Then probably both #588 and this issue should be handled as part of #382. Maybe you can add the things which you consider bugs to your comment in #382? Then we could close the new issues. Or we keep them open and simply link them in #382.

**Comment by sebastianpos** [comment]
Yes, please link them.

**Comment by zickgraf** [comment]
Fixed by c5ff6260c9dcafe818e4818a39dcdbfe3a729adf: The second argument of a final derivation must now always be a list of pairs (or triples, if other categories are specified) because I did not want to distinguish the case of bundled and non-bundled derivations. A warning is displayed if the given list of preconditions does not match the automatically detected one.

The problem that `WeakCokernelObject` etc. are not a bundled derivation is fixed in ddc38f04c3d6248991e966db839583d7e4322fdd.
